core_v2/sources/domains.move

```diff
@@ -517,6 +528,130 @@ module aptos_names_v2::domains {
         record.target_address = option::none();
     }
 
+    public fun renew_domain(
+        sign: &signer,
+        domain_name: String,
+        num_years: u8,
+    ) acquires CollectionCapabilityV2, NameRecordV2, RegisterNameEventsV1 {
+        // check the domain eligibility
+        let (_ , length) = validate_domain_name_string(domain_name);
+
+        assert!(!is_domain_in_renewal_window(domain_name), error::invalid_state(EDOMAIN_NOT_AVAILABLE_TO_RENEW));
+        let price = price_model::price_for_domain_v1(length, num_years);
+        renew_domain_internal(sign, domain_name, num_years, price);
+    }
+
+    fun renew_domain_internal(
+        sign: &signer,
+        domain_name: String,
+        num_years: u8,
+        price: u64,
+    ) acquires CollectionCapabilityV2, NameRecordV2, RegisterNameEventsV1 {
+        let record = get_record_mut(domain_name, option::none());
+        let registration_duration_secs = time_helper::years_to_seconds((num_years as u64));
+        record.expiration_time_sec = timestamp::now_seconds() + registration_duration_secs;
+
+        // pay the price
+        coin::transfer<AptosCoin>(sign, config::fund_destination_address(), price);
+        // log the event
+        event::emit_event<RegisterNameEventV1>(
+            &mut borrow_global_mut<RegisterNameEventsV1>(@aptos_names_v2).register_name_events,
+            RegisterNameEventV1 {
+                domain_name,
+                subdomain_name: option::none(),
+                registration_fee_octas: price,
+                expiration_time_secs: record.expiration_time_sec,
+            },
+        );
+    }
+
+    fun validate_domain_name_string(
+        domain_name: String,
+    ): (bool, u64) {
+        let (is_valid, length) = utf8_utils::string_is_allowed(&domain_name);
+        assert!(is_valid, error::invalid_argument(EDOMAIN_HAS_INVALID_CHARACTERS));
+        assert!(length <= config::max_domain_length(), error::out_of_range(EDOMAIN_TOO_LONG));
+        assert!(length >= config::min_domain_length(), error::out_of_range(EDOMAIN_TOO_SHORT));
+        return (is_valid, length)
+    }
+
+    public fun is_domain_in_renewal_window(
+        domain_name: String,
+    ): bool acquires CollectionCapabilityV2, NameRecordV2 {
+        // check if the domain is registered
+        assert!(name_is_registered(option::none(), domain_name), error::not_found(ENAME_NOT_EXIST));
+        // check if the domain is expired already
+        assert!(!name_is_expired(option::none(), domain_name), error::invalid_state(ENAME_NOT_AVAILABLE));
+        let record = get_record_mut(domain_name, option::none());
+
+        record.expiration_time_sec > timestamp::now_seconds() + MAX_REMAINING_TIME_FOR_RENEWAL_SEC
+    }
+
+    /// this is for domain owner to update subdomain expiration time
+    public fun renew_subdomain_as_domain_owner(
```

Comment:
since `renew_subdomain_as_domain_owner` is an entry function called directly by external parties, we should do some permission check, right? like checking signer is the subdomain's domain owner